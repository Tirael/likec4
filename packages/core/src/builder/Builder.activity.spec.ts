import { describe, expect, it } from 'vitest'
import { Builder } from './Builder'

describe('Builder (activity)', () => {
  const spec = Builder
    .specification({
      elements: {
        system: {},
        actor: {},
        component: {},
      },
    })

  it('should build ', () => {
    const b = spec.clone()
      .model(({ actor, system, component, activity }, _) =>
        _(
          actor('customer').with(
            activity('A'),
          ),
          system('cloud').with(
            component('ui').with(
              activity('B'),
            ),
          ),
        )
      )

    expect(b.build()).toMatchSnapshot()
  })

  it('should build steps 1', () => {
    const b = spec.clone()
      .model(({ actor, system, component, activity, step }, _) =>
        _(
          actor('customer').with(
            // activity('A'),
          ),
          system('cloud').with(
            component('api').with(
              activity('B', [
                '-> customer',
              ]),
            ),
            component('ui').with(
              activity('C', [
                step('-> cloud.api#B'),
                step('<- cloud.api#B'),
              ]),
            ),
          ),
        )
      )
    expect(b.build()).toMatchSnapshot()
  })

  it('should fail same hierarchy', () => {
    expect(() => {
      const b = spec.clone()
        .model(({ system, component, activity, step }, _) =>
          _(
            system('cloud').with(
              component('api').with(
                activity('B', {
                  steps: [
                    step('<- cloud.api'),
                  ],
                }),
              ),
            ),
          )
        )
    }).toThrowError('Invalid activity step between elements in the same hierarchy')
  })

  it('should build steps 2', () => {
    const b = spec.clone()
      .model(({ actor, system, component, activity, step }, _) =>
        _(
          actor('customer'),
          system('cloud').with(
            component('api'),
          ),
          activity('cloud.api#B'),
          component('cloud.ui').with(
            activity('C', [
              step('-> cloud.api#B'),
              step('<- cloud.api#B'),
            ]),
          ),
        )
      )
    expect(b.build()).toMatchSnapshot()
  })

  // it('should fail if invalid ID provided ', () => {
  //   expect(() => {
  //     spec.model(({ actor }, _) =>
  //       _(
  //         actor('cust.omer'),
  //       )
  //     )
  //   }).toThrowError('Parent element with id "cust" not found for element with id "cust.omer"')
  // })

  // it('should fail on invalid instance ', () => {
  //   const b = spec.clone()
  //     .model(_ =>
  //       _.model(
  //         _.component('cloud'),
  //         _.component('cloud.ui'),
  //       )
  //     )

  //   expect(() => {
  //     const raw = b.deployment(_ =>
  //       _.deployment(
  //         _.instanceOf('cloud.ui'),
  //       )
  //     ).build()
  //   }).toThrowError('Instance ui of cloud.ui must be deployed under a parent node')

  //   // Nested instanceOf is correct

  //   const raw = b.deployment(_ =>
  //     _.deployment(
  //       _.node('node').with(
  //         _.instanceOf('cloud.ui'),
  //       ),
  //     )
  //   ).build()
  //   expect(raw.deployments.elements).toEqual({
  //     node: expect.objectContaining({ id: 'node' }),
  //     // Take name from element
  //     'node.ui': expect.objectContaining({ id: 'node.ui', element: 'cloud.ui' }),
  //   })
  // })
})
