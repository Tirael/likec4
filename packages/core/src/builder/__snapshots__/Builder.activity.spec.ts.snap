// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Builder (activity) > should build  1`] = `
{
  "activities": {
    "cloud.ui#B": {
      "id": "cloud.ui#B",
      "modelRef": "cloud.ui",
      "name": "B",
      "steps": [],
    },
    "customer#A": {
      "id": "customer#A",
      "modelRef": "customer",
      "name": "A",
      "steps": [],
    },
  },
  "deployments": {
    "elements": {},
    "relations": {},
  },
  "elements": {
    "cloud": {
      "color": "primary",
      "description": null,
      "id": "cloud",
      "kind": "system",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "cloud",
    },
    "cloud.ui": {
      "color": "primary",
      "description": null,
      "id": "cloud.ui",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "ui",
    },
    "customer": {
      "color": "primary",
      "description": null,
      "id": "customer",
      "kind": "actor",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "imports": {},
  "relations": {},
  "specification": {
    "deployments": {},
    "elements": {
      "actor": {},
      "component": {},
      "system": {},
    },
    "relationships": {},
    "tags": [],
  },
  "views": {},
}
`;

exports[`Builder (activity) > should build steps 1`] = `
{
  "activities": {
    "cloud.api#B": {
      "id": "cloud.api#B",
      "modelRef": "cloud.api",
      "name": "B",
      "steps": [
        {
          "id": "step_cloud.api#B_1",
          "isBackward": false,
          "target": {
            "model": "customer",
          },
        },
      ],
    },
    "cloud.ui#C": {
      "id": "cloud.ui#C",
      "modelRef": "cloud.ui",
      "name": "C",
      "steps": [
        {
          "id": "step_cloud.ui#C_1",
          "isBackward": false,
          "target": {
            "activity": "cloud.api#B",
          },
        },
        {
          "id": "step_cloud.ui#C_2",
          "isBackward": true,
          "target": {
            "activity": "cloud.api#B",
          },
        },
      ],
    },
  },
  "deployments": {
    "elements": {},
    "relations": {},
  },
  "elements": {
    "cloud": {
      "color": "primary",
      "description": null,
      "id": "cloud",
      "kind": "system",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "cloud",
    },
    "cloud.api": {
      "color": "primary",
      "description": null,
      "id": "cloud.api",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "api",
    },
    "cloud.ui": {
      "color": "primary",
      "description": null,
      "id": "cloud.ui",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "ui",
    },
    "customer": {
      "color": "primary",
      "description": null,
      "id": "customer",
      "kind": "actor",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "imports": {},
  "relations": {},
  "specification": {
    "deployments": {},
    "elements": {
      "actor": {},
      "component": {},
      "system": {},
    },
    "relationships": {},
    "tags": [],
  },
  "views": {},
}
`;
